### PR TITLE
CsvInput verifies id spaces before starting import

### PR DESCRIPTION
**Pull request by tinwelint**
such that it will fail fast if there are relationship headers specifying
id spaces which no node header has will fail explaining just that instead of
treating all those relationships as bad.

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/Groups.java [hunk 1]
```diff
@@ -66,4 +66,36 @@ public synchronized Group getOrCreate( String name )
...
+    private IllegalStateException mixingOfGroupModesException()
+    {
+        return new IllegalStateException( "Mixing specified and unspecified group belongings " +
+                "in a single import isn't supported" );
+    }
+
+    public synchronized Group get( String name )
+    {
+        boolean global = name == null;
+        if ( globalMode != null && global != globalMode.booleanValue() )
```
Comment (klaren) [inline community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/Groups.java]: `.booleanValue()` is not needed

File: community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInput.java [hunk 2]
```diff
@@ -90,17 +98,74 @@ public CsvInput(
...
+                {
+                    Header header = relationshipHeaderFactory.create( dataStream, config, idType );
+                    verifyRelationshipHeader( header, Type.START_ID, dataStream.sourceDescription() );
+                    verifyRelationshipHeader( header, Type.END_ID, dataStream.sourceDescription() );
+                }
+            }
+        }
+        catch ( IOException e )
+        {
+            throw new UncheckedIOException( e );
```
Comment (klaren) [inline community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInput.java:144]: :+1:

File: community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/GroupsTest.java [hunk 3]
```diff
@@ -50,4 +52,109 @@ public void shouldHandleConcurrentGetOrCreate() throws Throwable
...
     }
+
+    @Test
+    public void shouldFailOnMixedGroupModeInGetOrCreate() throws Exception
+    {
+        // given
+        Groups groups = new Groups();
+        groups.getOrCreate( null );
+
+        try
```
Comment (klaren) [inline community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/GroupsTest.java]: Would prefer to have expected exceptions in the added tests since nothing is invoked after the throw.

File: community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputTest.java [hunk 4]
```diff
@@ -872,6 +885,95 @@ public void shouldSkipRelationshipValidationIfToldTo() throws Exception
...
+        try
+        {
+            // when
+            new CsvInput( data, defaultFormatNodeFileHeader(), relationshipData(), defaultFormatRelationshipFileHeader(), IdType.INTEGER,
+                    COMMAS, mock( Collector.class ), getRuntime().availableProcessors(), false );
+            fail( "Should not parse" );
+        }
+        catch ( InputException e )
+        {
+            // then
```
Comment (klaren) [inline community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputTest.java:903]: Same here

File: community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputTest.java [hunk 5]
```diff
@@ -1005,6 +1107,51 @@ private static CharReadable charReader( String data )
...
+        CapturingDataFactories( Function<Configuration,CharReadable> factory, Decorator<ENTITY> decorator )
+        {
+            this.factory = factory;
+            this.decorator = decorator;
+        }
+
+        @Override
+        public Iterator<DataFactory<ENTITY>> iterator()
+        {
+            return Iterators.iterator( new DataFactory<ENTITY>()
```
Comment (klaren) [inline community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputTest.java]: Great place for a lambda ;)
```java
          return Iterators.iterator( config -> new Data<ENTITY>()
            {
                @Override
                public CharReadable stream()
                {
                    return last = factory.apply( config );
                }

                @Override
                public Decorator<ENTITY> decorator()
                {
                    return decorator;
                }
            } );
```

File: community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/Groups.java [hunk 6]
```diff
@@ -66,4 +66,36 @@ public synchronized Group getOrCreate( String name )
...
         return group;
     }
+
+    private IllegalStateException mixingOfGroupModesException()
+    {
+        return new IllegalStateException( "Mixing specified and unspecified group belongings " +
+                "in a single import isn't supported" );
+    }
+
+    public synchronized Group get( String name )
```
Comment (klaren) [inline community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/Groups.java:76]: It's a bit hard to follow the flow, but it looks like this method was introduced to help `verifyRelationshipHeader()` assert that the header is there. Wouldn't a `Groups.isPresent(String)` make more sense and eliminate the duplicate code?

**Review by klaren — CHANGES_REQUESTED** [review]
Fail fast and early! :tada:

I just have a few small comments.

File: community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/Groups.java (hunk 6 above)
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/Groups.java:76]: `isPresent` would be implemented almost like `get` wouldn't it?

Comment (klaren) [reply community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/Groups.java:76]: That is the case, yes, disregard my previous comment :)

**Review by klaren — APPROVED** [review]
